[1.0.0-alpha2] Workload and component (#316)
oam-dev/spec: 3.workload.md

A core workload schema MUST be in the `core.oam.dev` group. A core workload MUST all be supported according to the definitions in the spec by any runtime implementation of this specification. Here is an example of a core workload definition.

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: x.core.oam.dev
spec:
  definitionRef:
    name: x.core.oam.dev
```

Currently, all core workloads are container-based, and assume that an implementation is capable of executing an OCI or Docker image as a container, and are capable of working with OCI registries.
We may introduce non container-based workloads, such as VM-based, in the future. We can also promote standard workloads to core if there is a consensus in the community.

### Standard Workload

A standard workload schema MUST be in the `standard.oam.dev` group. Here is an example of a standard workload definition.

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: x.standard.oam.dev
spec:
  definitionRef:
    name: x.standard.oam.dev
```

A standard workload MUST be implemented according to the definitions in the spec if any runtime chooses to implement it.
 

> We do not have any standard workload defined in the current spec. We may introduce new workloads or promote extended workloads to standard if there is a consensus in the community.

Reviewer: Can we make a clear claim that standard workloads will be promoted from extended workloads? I'm hoping users can get the point of "promotion ladder" in this section.

Author: I am not sure if the standard workloads have to be promoted from an extended one. It can be a new one that is similar to a bunch of extended workloads. The core one has to be promoted from standard as we won't change its definition.

Reviewer: Emm, can we highlight the idea of promotion ladder somewhere? It seems implicit for now.

# 3. Workload

> 

Reviewer: Maybe we can add a `**Target User: Infrastructure Operator**` section here (and to the beginning of every other doc)?

This section defines the workload definition schematics.

Workload definitions are provided by the platform (just as traits are) so that users may inspect the platform and
 learn what workloads are available for use. Workload definitions are not extensible to platform users (only to platform

>   operators). Thus, platform users MUST NOT be allowed to create new workload kinds.

Reviewer: We need to use terms like `Application Developers` etc in the whole doc instead of "platform users" etc for better consistency.

Another suggestion to make above narrative clearer:

Workload Definition is provided by the platform to enforce and validate the workload specification of given `Component`. It's designed to be a platform level concept and expected to be managed by Infrastructure Operators only. The OAM platform should provide a proper approach (e.g. docs or samples) to help Application Developers define `spec.workload` section in Component instead of exposing Workload Definitions directly to them.

Author: > Put otherwise, I agree that we should explain that `WorkloadDefinition` is intended for use by infrastructure operators, but I do not think we should go so far as to say application operators MUST NOT be allowed to create workload kinds.

Agree,  this is copied from the existing spec (https://github.com/oam-dev/spec/blob/master/7.workload_types.md#7-workload-types) so I didn't change it as this PR is just about the schema change.  I didn't know why we put such a strong statement in the begining. We can add a PR to resolve this.

Reviewer: I'd suggest we change the narrative in this PR :-)

## Top-Level Attributes

> The following attributes are common across all schemata defined in this document. This structure is designed to allow

Reviewer: I hope we can delete this section from every doc, and link/copy Kubernetes resource model documentation here (https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/#required-fields).

Author: This only happens once.  Also, I thought  that our GKV are "inspired" by the K8s but  not direct dependent on K8s definition.

Reviewer 2: Strongly agree. :) Though we might consider linking instead to the API conventions document; I believe this is the closest thing that exists to a formal specification of the Kubernetes API.

https://github.com/kubernetes/community/blob/1d6d124/contributors/devel/sig-architecture/api-conventions.md

Reviewer: Yes, the api-conventions.md doc is the best candidate here. It also reminds me that referring to an epoch design doc of Kube feels much better than linking K8s doc. Good point!

To @ryanzhang-oss , for metadata and resource model, I'd suggest we directly use it instead of "inspired". "Inspire" is not good unless we significantly improved it and decide not contribute back. 

If needed (seems not IMO), we can clearly clarify that we only support a certain subset of the whole k8s resource model convention. But I'd strongly prefer eliminating the feeling of "inspire".

 object type to be determined by scanning two fields (`apiVersion` and `kind`), and provide access to common fields
  in the `metadata` section. The `spec` section is where type-specific information is located.

Here are the attributes that provide top-level information about the workload definition.

| Attribute | Type | Required | Default Value | Description |
|-----------|------|----------|---------------|-------------|
| `apiVersion` | `string` | Y | | The specific version of the specification in use. The core types use `core.oam.dev/v1alpha2` |
| `kind` | `string` | Y || The string `WorkloadDefinition` |
| `metadata` | [`Metadata`](#metadata) | Y | | WorkloadDefinition metadata. |
| `spec`| [`Spec`](#spec) | Y | | A container for the workload definition. |

### Metadata

> The metadata describes this workload definition. The `metadata` section is defined in [the meta section of part 2](2.overview_and_terminology.md#Metadata)

Reviewer: I also hope we can directly reference `metadata` from Kubernetes resource model (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#objectmeta-v1-meta) in "overview and terminology" doc. 

I'm not in favor of making the spec tedious and complex due to explanations on resource model and metadata.

Author: Same as above. Btw, this section is copied from the existing spec

Reviewer: This doc is better: https://github.com/kubernetes/community/blob/1d6d124/contributors/devel/sig-architecture/api-conventions.md#metadata

My opinion is we need to have a clearer relationship between overview_terminology.md section and K8s resource model than today. /cc @negz 

We can fix this in following up PR of course.

Reviewer 3: I'm hesitant about linking directly to the Kubernetes docs for formal definitions, including structure or field names. If Kubernetes decides to suddenly change then name or usage of a field, we would need to first assess that change and decide if it should be applied here too. Conceptual docs are probably fine since a change to the wording of a concept in the Kubernetes doc won't necessarily break an OAM runtime.

Reviewer: @vturecek Actually, `.../generated/kubernetes-api/{version}` is the permanent link of versioned K8s api spec. It's much safer than what OAM spec currently linked in the doc, e.g.: https://github.com/oam-dev/spec/blob/master/2.overview_and_terminology.md#label-format which only reflects the latest version of K8s api.

Also, I propose we directly refer K8s metadata instead of creating a "simplified" version, because in practice, the OAM K8s runtime still need to use full version of [k8s metadata](https://github.com/oam-dev/rudr/blob/319448fb4cfd970152ecb24d3722536d66ea6bd7/examples/components.yaml#L3). The OAM spec version of "name + labels + annotations" actually is not enough to work at all. In that sense, current rudr doesn't fully follow OAM spec.

I'm totally OK we fix it in following PR but we need to fix it ASAP indeed.

Reviewer: What I proposed above also applies to GVK section and many places in "overview and terminology" doc . Non-k8s impl will also be benefit a lot since it could import K8s types directly in many languages. It doesn't make sense to create sth "similar" in OAM spec.

Reviewer 3: So are we saying that OAM-compliant runtimes must implement the full Kubernetes metadata object model, as specified in the Kubernetes API conventions doc? That seems extreme to me. Namespaces would be required by every runtime even though that has never been a concept we've had in OAM and to me seems outside of the scope of an application modeling definition.

As I stated earlier, I'm fine linking off to _concepts_ where we feel they are identical. If the content is versioned, all the better. 

I liked the suggestion earlier that we clearly label parts of the spec with the role they are most relevant to (app developer, app operator, infra operator). But I'm not sure what it means to have a separate part of the spec for platform builders, or what we gain from that. If they are split then I'm assuming there are object models we define for platform builders that end users don't see, in which case I'm not sure why we even care about those.

Reviewer: @vturecek I also agree directly referring Kubernetes design doc is a good start to make the spec doc simpler.

While after digesting your comment above, I feel that the OAM spec doc cares very much about end users. But honestly, my personal observation is the current spec is barely friendly to them. I'd guess it's also the main reason that OAM spec was commented as "complex":  it reads like a doc for someone to implement an OAM platform rather than material to grape OAM concepts immediately at first glance, or more possibly, those two are inextricably mixed together in the spec.

In that sense, it's exactly the same mistake that K8s Deployment is making and we claim to solve in OAM :-)

Another concern is, the current spec is not implementers friendly either. For example, from reading the spec, I have feeling that we are suggesting implementers to code OAM metadata types from scratch. I really doubt there's any implementer will do that. /cc @negz 

So I don't mind if it's named "splitting", but I'd highlight the idea like: OAM spec **should also consider separate concerns** in its content, specifically:

1. straightforward docs for end users to get the core ideas of OAM in 10 min instead of reading a 10 pages specification. IMO, detailed fields of OAM metadata is not part of this. 
2. detailed and accurate docs for implementers to implement OAM on whatever runtime, and what's more important, with minimal effort. For example, again the `metadata` section, the message will be much clearer like "OAM metadata is a subset of K8s metadata (link) and here're the fields OAM cares".

I don't see any reason that we can't achieve this or have to mix them in one spec doc. Though I'd suggest we move this discussion out of this PR, it need to be fixed by following ups (and would be better to involve a tech writer) /cc @ryanzhang-oss

### Spec

> The specification section contains a reference to the workload schema

Reviewer: ```suggestion
The specification section contains a reference of the workload's definition.
```

| Attribute | Type | Required | Default Value | Description |
|-----------|------|----------|---------------|-------------|
| `DefinitionRef` | [`DefinitionRef`](#DefinitionRef) | Y | | Workload schema reference. |

#### DefinitionRef

> DefinitionRef defines where to find the workload schema

Reviewer: ```suggestion
DefinitionRef defines the schema of workload. Note that it could be simply an indexing for a schema.
```

Reviewer 2: > Note that it could be simply an indexing for a schema.

It's not immediately obvious to me what this is intending to communicate. Could we word it more clearly?

| Attribute | Type | Required | Default Value | Description |
|-----------|------|----------|---------------|-------------|

> | `name` | `string` | Y | | The workload schema's name. Must be formatted in a way that can uniquely identify a specific version of the definition  |

Reviewer: We really need sample to make spec easy to understand.

Also I'm not sure that "indexing" works for non-CRD schema. I'm thinking maybe:

```yaml
definitionRef:
  name: scaling.alibaba.com/v1alpha2.Manualscaling
  schematic: "json schema string" # this is a optional field only needed for non-CRD platforms
```

Reviewer 2: > Must be formatted in a way that can uniquely identify a specific version of the definition

Cross-linking this discussion; https://github.com/crossplane/crossplane/pull/1276#pullrequestreview-362313530. I'm really not sure requiring a specific version makes sense in the Kubernetes world.

Author: would it make sense in a non-k8s world?

Reviewer: I am in favor of Ryan's proposal:

```yaml
name: containerisedworkload.core.oam.dev
version: v1beta1
```
The version field should be optional of course (only needed when multiple version CRDs are installed which is a rare case).

Author: > Also I'm not sure that "indexing" works for non-CRD schema. I'm thinking maybe:

it seems that it should work as long as the "name" is unique? Is it hard to do? With the schematic and name together, what does the "name" mean here?

The workload definition itself __MUST__ contain information that can be used to locate it directly. We will use that
 information in the [Component](4.component.md) to refer to the workload schema instead of using the name of the workload definition. 

> We recommend that the `name` here follows the format described in [the Group, Version, Kind section of part 2](2.overview_and_terminology.md).

Reviewer 2: Is this suggesting, for example, `name: local.dev/v7alpha2.Proxy`? I recommend we do not use a field named `name` unless we're actually referring to the `metadata.name` of some other Kubernetes object, and I don't believe this is a valid Kubernetes object name (per https://kubernetes.io/docs/concepts/overview/working-with-objects/names/).

My thinking here was that `name` would refer to a named instance of a `CustomResourceDefinition`. If we instead want to simply refer to a particular kind of resource we should use `apiVersion` and `kind` (and not `name`).

Reviewer: I think it's fine to have `name` (and it's name of CRD) in xxxDefinition, but in other obj, I agree with your point.

Reviewer 4: > I think it's fine to have name (and it's name of CRD) in xxxDefinition, but in other obj, I agree with your point.

it seems DefinitionRef Name here is a GVK string. Yes we can find the referred CR instance by parsing it but the combination of name: GVK as Nic suggested, is not very intuitive

> We also recommend that the `name` of the workload (in the metadata section) to include the Group and Kind of the schema.

Reviewer: I'd strongly suggest to include samples instead of jumping to some other place. The current narrative is not easy to understand even from my view.

> ## Workload Kinds

Reviewer: Kinds have special meaning in K8s world. Let's name it "Workload Ladder" or sth fancy.

There are three kinds of workloads:

__Core workloads__ are the kind of workloads that every OAM runtime MUST implement according to spec definition

__Standard workloads__ are the kind of workloads that a OAM runtime can choose to implement. However, the
 implementation MUST strictly adhere to the schematics defined in the spec 

> __Extended workloads__ OAM platforms are free to define extra workload kinds and  

Reviewer: and ... ?

### Core Workload Kind

A core workload MUST be in the `core.oam.dev` group.

A core workload MUST all be supported according to the definitions in this section by any runtime implementation of
 this specification.

Currently, all core workloads are container-based, and assume that an implementation is capable of executing an OCI or Docker image as a container, and are capable of working with OCI registries.
We may introduce non container-based workloads, such as VM-based, in the future. We can also promote standard workloads to core if there is a consensus in the community.

### Standard Workload Kind

A standard workload MUST be in the `standard.oam.dev` group.

A standard workload MUST be implemented according to the definitions in this section if any runtime chooses to implement it.
 
We do not have any standard workload defined in the current spec. We may introduce new workloads or promote extended workloads to standard if there is a consensus in the community.
 
### Extended Workload Kind

> An extended workload MUST be in the `extended.oam.dev` group.

Reviewer 2: Part of our goal with this spec update was to allow arbitrary Kubernetes CRs to act as extended workloads. This isn't possible if all extended workloads MUST be in the `extended.oam.dev` API group. I don't think we should have an opinion about the group name of extended workloads.

Author: >  This isn't possible if all extended workloads MUST be in the `extended.oam.dev` API group.

Since the CRDs' group name does not have to be the same as its parent workloadDefinition's group name, I wonder why it is not possible to refer to any CRDs?

Reviewer 2: Perhaps I'm misunderstanding something here. This reads to me like you're saying any extended workload, i.e. any kind blessed as a workload via a `WorkloadDefinition` must be in the `extended.oam.dev` group. So for example I could not use a Crossplane `RedisCluster` as a workload, because it's in `cache.crossplane.io`, not `extended.oam.dev`:

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: Component
metadata:
  name: example-cache
spec:
  workload:
    # Are you saying this MUST be extended.oam.dev?
    apiVersion: cache.crossplane.io/v1alpha1
    kind: RedisCluster
    spec:
      engineVersion: "3.2"
```

Reviewer: @negz  I think Ryan means the WorkloadDefinition under `extended.oam.dev` group instead of `core.oam.dev`? It should be mentioned in xxxDefinition section.

Reviewer 5: Yeah, the description here has some mislead.  It is workloadDefinition MUST be in the `extended.oam.dev` group not workload itself.

Reviewer 2: Does this mean we'll have multiple `kind: WorkloadDefinition`, each in a different API group (`extended.oam.dev`, `core.oam.dev`, etc?). I imagine core workloads will still require a `WorkloadDefinition`.

Reviewer 5: Sorry @negz ,  I'm wrong.  I think we didn't need `extended.oam.dev` group at all.  I think the group `core.oam.dev` means they are absolutely defined in OAM Spec and every OAM runtime MUST follow it.  The group `standard.oam.dev` means they are defined in OAM Spec and can be optionally implemented by runtime.

Reviewer 3: This also contradicts the recommendation above, which says that we recommend the workload definition name match the schema name. For extended workload types, that will be impossible if someone is adding existing workloads, like the RedisCluster in @negz 's example. 

I don't think we need the `extended.oam.dev` group at all here.

Author: > ```yaml
>   workload:
>     # Are you saying this MUST be extended.oam.dev?
>     apiVersion: cache.crossplane.io/v1alpha1
>     kind: RedisCluster
>     spec:
>       engineVersion: "3.2"
> ```
No, this is the apiVersion of the CRD. The extended workloadDefinition itself needs to be in the "extended.oam.dev". The CRD it refers to can be anything.

Reviewer 4: > workload:
>     # Are you saying this MUST be extended.oam.dev?
>     apiVersion: cache.crossplane.io/v1alpha1
>     kind: RedisCluster
>     spec:
>       engineVersion: "3.2"
> 
> No, this is the apiVersion of the CRD. The extended workloadDefinition itself needs to be in the "extended.oam.dev". The CRD it refers to can be anything.

meaning
```
apiVersion: **extended**.oam.dev/v1alpha1
kind: WorkloadDefinition
metadata:
  name: rediscluster.cache.crossplane.io
spec:
  definitionRef:
    name: rediscluster.cache.crossplane.io
```
?

Author: > apiVersion: **extended**.oam.dev/v1alpha1
> kind: WorkloadDefinition
> metadata:
>   name: rediscluster.cache.crossplane.io
> spec:
>   definitionRef:
>     name: rediscluster.cache.crossplane.io

yeah,  is this ok?

Reviewer 6: > allow arbitrary Kubernetes CRs to act as extended workloads.

I totally agree on this. OAM is supposed to manage, not replace CRDs. The "extended" needs some context that needs some explanation:

When we started the project, our vision is to standardize application management. There are levels of standardization in our roadmap: _core_ for minimum, _standard_ for cloud ubiquitous, and _extended_ for things that are popular but not ubiquitous. We host and govern those features/CRDs/Operators in OAM to build a standardized application delivery experience.

This is not a technical issue, but more a community building effort.

Reviewer: Allow me summarize a bit, so Ryan proposed is:

For core workloads, it looks like:
```yaml
apiVersion: core.oam.dev/v1alpha1
kind: WorkloadDefinition
metadata:
  name: x
spec:
  definitionRef:
    name: x.core.oam.dev
```
For standard workload, it will be:

```yaml
apiVersion: standard.oam.dev/v1alpha1
kind: WorkloadDefinition
metadata:
  name: y.database.alibaba.com
spec:
  definitionRef:
    name: y.database.alibaba.com
```

For extended workload, it will be:

```yaml
apiVersion: extended.oam.dev/v1alpha1
kind: WorkloadDefinition
metadata:
  name: z.cache.crossplane.io
spec:
  definitionRef:
    name: z.cache.crossplane.io
```
Am I right?

Author: Here is what I think makes sense.  WorkloadDefinition always belongs to core.oam.dev as it's OAM's core concept. The apiground in the schema for core workload is core.oam.dev.  The apiground in the schema for standard workload is standard.oam.dev. There is no restriction on the api-group for extended.

For core workloads, it looks like:
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: x.core.oam.dev
spec:
  definitionRef:
    name: x.core.oam.dev
```
For standard workload, it will be:
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: y.standard.oam.dev
spec:
  definitionRef:
    name: y.standard.oam.dev
```
For extended workload, it will be:
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: z.cache.crossplane.io
spec:
  definitionRef:
    name: z.cache.crossplane.io
```

Reviewer 2: I think I had a stale view of this thread when I wrote https://github.com/oam-dev/spec/pull/316#discussion_r384223831. It sounds like I agree with @ryanzhang-oss's comment at https://github.com/oam-dev/spec/pull/316#discussion_r384219536.

Extended workload kind are _per runtime_, meaning that each runtime may define its own extended workloads schema beyond this specification. 

## Core Workload
The following core workload kinds are defined by this specification

> ### Container Workload

Reviewer: I strongly suggest to move spec of core workloads (the whole section below) to a separate directory (e.g. `spec/core_workloads/container_workload.md` etc), and we will have more core workloads in the future.

My opinion is they are not part of OAM spec, they just allowed to stay in OAM spec repo because they are "core".

Author: I thought about that too, didn't know where to put them.

Reviewer: Let's do `spec/core_workloads/container_workload.md`. /cc @negz

Reviewer 2: Happy for it to move out of this doc. I suggest calling it a "_containerised_ workload" rather than a "container workload" though.

Author: `containerized` or `containerised`?

Reviewer 2: This is very hard to say an an Australian, but since software tends to use US english it should probably be containeri**z**ed. 😢

A containerized workload is used for long-running workloads. Here is the definition of the containerized workload

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: workloadDefinition
metadata:
  name: containerWorkload.core.oam.dev
spec:
  definitionRef:

>     name: core.oam.dev/v1alpha2.containerWorkload

Reviewer 2: I believe `core.oam.dev/v1alpha2.ContainerWorkload` is not a valid `metadata.name` (due to the `/` and the uppercase characters) and thus should not be the value of a field named `name`. If what we're trying to do is reference a particular GVK (not a named instance of, for example, a CRD) then we should write:

```yaml
definitionRef:
  apiVersion: core.oam.dev/v1alpha2
  kind: ContainerisedWorkload
```

# 3. Workload

This section defines the workload definition schematics.

Usually, workload definitions are provided by the OAM runtime (just as traits are) so that application developers may inspect the OAM runtime and learn what workloads are available for use.  

## Top-Level Attributes

> The following attributes are common across all schemata defined in this document. They follow the [Kubernetes specification](https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/#required-fields).

Author: I think MSFT/Alibaba/Upbound should find time this week to discuss this. I think we should address this in a follow-up PR since this is only getting into alpha2 branch. This PR is more for the technical part.

| `metadata` | [`Metadata`](#metadata) | Y | | WorkloadDefinition metadata. |
| `spec`| [`Spec`](#spec) | Y | | A container for the workload definition. |

### Metadata

The metadata describes this workload definition. The `metadata` section is defined in [the meta section of part 2](2.overview_and_terminology.md#Metadata)

### Spec

The specification section contains a reference to the workload definition.

| Attribute | Type | Required | Default Value | Description |
|-----------|------|----------|---------------|-------------|
| `definitionRef` | [`DefinitionRef`](#DefinitionRef) | Y | | Workload schema reference. |

#### DefinitionRef

DefinitionRef defines an index to find the workload schema.

| Attribute | Type | Required | Default Value | Description |
|-----------|------|----------|---------------|-------------|
| `schemaName` | `string` | Y | | The workload schema's name. |

The workload schema definition itself __MUST__ contain information that can be used to uniquely identify it. OAM runtime will use that information in a [Component](4.component.md) to refer to the workload schema instead of using the name of the workload definition.

>  We recommend to use the format [Group/Version/Kind](2.overview_and_terminology.md#Group) described in section 2 as the unique identification in the schema. 

Author: > ```yaml
> spec:
>   definitionRef:
>     name: core.oam.dev/v1alpha1.ContainerizedWorkload
> ```
> 
> Is this recommending that runtimes use the above `group/version.Kind` syntax to uniquely identify a schema? I think this would be surprising and misleading to folks familiar with Kubernetes, and make implementing the spec a little more difficult for runtimes that are built using Kubernetes.
> 
> Quoting API conventions:
> 
> > The name of a field referring to another resource of kind Foo by name should be called fooName. The name of a field referring to another resource of kind Foo by ObjectReference (or subset thereof) should be called fooRef.
> 
> > Object references should either be called fooName if referring to an object of kind Foo by just the name (within the current namespace, if a namespaced resource), or should be called fooRef, and should contain a subset of the fields of the ObjectReference type.
> 
> i.e. `spec.fooRef.name` should refer to the `metadata.name` of a `kind: Foo`. In the case of `definitionRef` we omit the kind and leave it up to the runtime (perhaps it's a `kind: CustomResourceDefinition`, perhaps it's a `kind: SweetXMLSchema`) but for Kubernetes runtimes I think this should be the _name_ of the definition, not a GVK. The `group/version.Kind` syntax is not a valid Kubernetes name (due to the `/` and uppercase kind), and CRDs specifically [MUST](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition) be named plural.group, e.g. `containerizedworkloads.oam.core.dev`.
> 
> So TL;DR I expect all runtimes that are implemented atop the Kubernetes API server would need to write:
> 
> ```yaml
> spec:
>   definitionRef:
>     name: containerisedworkloads.oam.core.dev
> ```
No, that link actually points to exactly what you describe here

We recommend that the `schemaName` here follows the format described in [the Representations of Group/Version/Kind in section 2](2.overview_and_terminology.md#Representations).

We also recommend that the `name` of the workloadDefinition to be the same as the `schemaName` it refers to. Here is an example

```yaml
apiVersion: core.oam.dev/v1alpha2
kind: WorkloadDefinition
metadata:
  name: schema.example.oam.dev
spec:
  definitionRef:
    schemaName: schema.example.oam.dev

> ```

Reviewer 3: the example here doesn't follow the recommendation we just gave above to use GVK. Is it enough to simply recommend GVK, or should it be required?

Author: We didn't recommend GVK as the result of several discussions says no "V" needed. I will modify the wording a bit.

Reviewer 6: Hi @vturecek 

Please see referenced discussions 
- https://github.com/oam-dev/spec/pull/316#discussion_r383492681
- https://github.com/crossplane/crossplane/pull/1276#discussion_r382321173 (@negz 's PR in crossplane)

__Extended workloads__ are the kind of workloads that a OAM runtime is free to define.

### Core Workload

A core workload MUST be in the `core.oam.dev` group.

A core workload MUST all be supported according to the definitions in this section by any runtime implementation of this specification.

Currently, all core workloads are container-based, and assume that an implementation is capable of executing an OCI or Docker image as a container, and are capable of working with OCI registries.
We may introduce non container-based workloads, such as VM-based, in the future. We can also promote standard workloads to core if there is a consensus in the community.

### Standard Workload

A standard workload MUST be in the `standard.oam.dev` group.

A standard workload MUST be implemented according to the definitions in this section if any runtime chooses to implement it.
 
We do not have any standard workload defined in the current spec. We may introduce new workloads or promote extended workloads to standard if there is a consensus in the community.
 
### Extended Workload
An extended workload MUST be in the `extended.oam.dev` group.

> Extended workload kind are _per runtime_, meaning that each runtime may define its own extended workloads schema beyond this specification. 

Author: Agree,  I took a shoot at re-wording it,  not my forte...

## Core Workload definitions
The following core workloads are defined in our current OAM specification.

> ### Containerize Workload

Reviewer 2: ```suggestion
### Containerized Workload
```
